File: iosApp/iosApp/Login/Components/LegalContent.swift
import Foundation

struct LegalParagraph: Hashable {
    let text: String
    var highlights: [String] = []
    var isHeading: Bool = false
}

/// Parses headings and inline emphasis from the small Markdown subset used by Demo legal resources.
func parseLegalDocument(_ body: String) -> [LegalParagraph] {
    body.components(separatedBy: "\n\n").compactMap { raw in
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return nil }
        let isHeading = value.hasPrefix("## ")
        let marked = isHeading ? String(value.dropFirst(3)) : value
        let components = marked.components(separatedBy: "**")
        let highlights = components.enumerated().compactMap { index, part in
            index.isMultiple(of: 2) || part.isEmpty ? nil : part
        }
        return LegalParagraph(
            text: components.joined(),
            highlights: highlights,
            isHeading: isHeading
        )
    }
}

File: androidApp/src/main/kotlin/com/example/demo/login/components/LegalContent.kt
package com.example.demo.login.components

data class LegalParagraph(
    val text: String,
    val highlights: List<String> = emptyList(),
    val isHeading: Boolean = false
)

private val LegalHighlightPattern = Regex("\\*\\*(.+?)\\*\\*")

/** Parses the small Markdown subset used by the bundled Demo legal documents. */
fun parseLegalDocument(body: String): List<LegalParagraph> = body
    .split(Regex("\\n\\s*\\n"))
    .mapNotNull { raw ->
        val value = raw.trim()
        if (value.isEmpty()) return@mapNotNull null
        val isHeading = value.startsWith("## ")
        val markedText = value.removePrefix("## ")
        val highlights = LegalHighlightPattern.findAll(markedText).map { it.groupValues[1] }.toList()
        LegalParagraph(
            text = markedText.replace(LegalHighlightPattern, "$1"),
            highlights = highlights,
            isHeading = isHeading
        )
    }
